include/revng/StackAnalysis/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1089 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Binary.h"
+#include "revng/Model/Register.h"
+#include "revng/Model/Type.h"
+
+namespace abi {
+
+/// A trait struct which is specialized for all of the supported ABIs.
+/// It describes the specifics of the calling convention for the ABI.
+template<model::abi::Values V>
+struct CallingConventionTrait {
+
+  /// States whether specified ABI allows using both general purpose and vector
+  /// registers in parallel, e.g. whether `GeneralPurposeArgumentRegisters[i]`
+  /// can be used for an argument in a functions that also uses
+  /// `VectorArgumentRegisters[i]` for a different argument when `i` satisfies
+  /// `i < GeneralPurposeArgumentRegisters.size()` and
+  /// `i < VectorArgumentRegisters.size()`.
+  ///
+  /// For an example of an ABI with position based registers see Microsoft x64
+  /// documentation, for one without - see SystemV x64 documentation.
+  static constexpr bool ArgumentsArePositionBased = false;
```

Comment:
The rabbithole goes way deeper than that. I guess, this comment fails to explain the situation after all.

The idea behind the argument is to differentiate ABIs that preserve argument ordering whatever the cost from those that don't care.
Here's an example:
```cpp
struct A; // sizeof(A) == 1. And it uses some special alignment, so it can only be passed in memory accordingly go C++ ABI.
void function(A, int, float);
```
A position based ABI would pass those like
- `A` in memory. (Maybe a pointer in the first GPR, depending on the exact ABI)
- `int` in the second GPR.
- `float` in the third vector/coprocessor register for floats.
A non-position-based ABI would do something like:
- `A` in memory.
- `int` in the first GPR.
- `float` in the first vector/coprocessor register for floats.

I hope, it's a clearer way to describe the situation. The name doesn't suit it, sure, but we shouldn't hide our head into the float-ish stuff. It affects the floats significantly, but that's not what the option is meant to represent in its core.

Reply:
I see. I definitely missed all of this. Now the name makes much more sense, but the description is heavily focused on vector registers.

An alternative point of view would be "GreedyArgumentAllocation", to convey that the ABI uses all the arguments it can as early as it can.

Otherwise, we can keep the name but switch the description to something closer to what you explained here.

Reply:
I think `Greedy` naming is even further of truth. Since, as far as I understand, the tradeoff here is whether ABI wants maximum performance or whether it wants to be able to reliably tell the actual order of the arguments. That's why I kept sneaking the `Position` word into the name.

Reply:
It's not clear to me if this is vector vs GPR or not. If it's only vector vs GPR, what about `VectorArgumentsIndicesEvolveAutonomously`?
Are you sure it's suitable to load this field with the behavior of big arguments that end up in the stack too?

Reply:
It also affects stack.

On that note, there's also multiple ways the ABI passes arguments (for example some have multiple ways for floating point types of different sizes: floating point stack for small stuff and new AES registers for bigger variables) that is also affected by this distinction.

Reply:
OK, it seems to me that, given the insights you got reading a lot of ABIs, you really think this is a key distinction with a significant expressive power that should not be split in multiple parts.
Then let's go with it!